src/algebra/quandle.lean

```diff
@@ -0,0 +1,561 @@
+Authors: Kyle Miller
+-/
+import algebra
+import data.zmod.basic
+import data.equiv.mul_add
+
+/-!
+# Racks and Quandles
+
+This file defines racks and quandles, algebraic structures for sets
+that bijectively act on themselves with a self-distributivity
+property.  If `R` is a rack and `op : R → (R ≃ R)` is the self-action,
+then the self-distributivity is, equivalently, that
+```
+op (op x y) = op x * op y * (op x)⁻¹
+```
+where multiplication is composition in `R ≃ R` as a group.
+Quandles are racks such that `op x x = x` for all `x`.
+
+One example of a quandle is the action of a Lie algebra on itself,
+defined by `op x y = Ad (exp x) y`.
+
+Quandles and racks were independently developed by multiple
+mathematicians.  David Joyce introduced quandles in his thesis
+[Joyce1982] to define an algebraic invariant of knot and link
+complements that is analogous to the fundamental group of the
+exterior, and he showed that the quandle associated to an oriented
+knot is invariant up to orientation-reversed mirror image.  Racks were
+used by Fenn and Rourke for framed codimension-2 knots and
+links.[FennRourke1992]
+
+The name "rack" came from wordplay by Conway and Wraith for the "wrack
+and ruin" of forgetting everything but the conjugation operation for a
+group.
+
+## Main definitions
+
+* `rack`
+* `quandle`
+* `quandle.conj` defines a quandle of a group acting on itself by conjugation.
+* `rack_hom` is homomorphisms of racks and quandles.
+* `rack.envel_group` gives the universal group the rack maps to as a conjugation quandle.
+* `rack.opp` gives the rack with the action replaced by its inverse.
+
+## Main statements
+* `rack.envel_group` is universal (`to_envel_group.univ` and `to_envel_group.univ_uniq`)
+
+## Notation
+* `x ◃ y` is local notation for `rack.op x y`
+* `x ◃⁻¹ y` is local notation for `rack.inv_op x y`
+
+## Todo
+
+* If g is the Lie algebra of a Lie group G, then (x ◃ y) = Ad (exp x) x forms a quandle
+* If X is a symmetric space, then each point has a corresponding involution that acts on X, forming a quandle.
+* Alexander quandle with `a ◃ b = t * b + (1 - t) * b`, with `a` and `b` elements of a module over Z[t,t⁻¹].
+* If G is a group, H a subgroup, and z in H, then there is a quandle `(G/H;z)` defined by
+  `yH ◃ Hx = yzy⁻¹xH`.  Every homogeneous quandle (i.e., a quandle Q whose automorphism group acts
+  transitively on Q as a set) is isomorphic to such a quandle.
+  There is a generalization to this arbitrary quandles in [Joyce's paper (Theorem 7.2)][Joyce1982].
+
+## Tags
+
+rack, quandle
+-/
+
+universes u v
+
+/--
+A *rack* is an automorphic set (a set with an action on itself) that is self-distributive.
+The `x ◃ y` and `x ◃⁻¹ y` notation is right associative.
+
+There is a weaker notion of a *shelf*, which would be from having `op' : α → α → α`.
+-/
+class rack (α : Type*) :=
+(op' : α → (α ≃ α))
+(self_distrib' : ∀ {x y z : α}, op' x (op' y z) = op' (op' x y) (op' x z))
```

Comment:
Are you sure you actually need the primed and unprimed versions here?

Reply:
I didn't look too much into what the problem was, but without the primed version none of the expressions were pretty printing with the local notation.

I wonder if I should expand out the `equiv` so that the definition is the following?
```lean
class rack (α : Type*) :=
(act : α → α → α)
(inv_act : α → α → α)
(left_inv : ∀ x, function.left_inverse (inv_act x) (act x))
(right_inv : ∀ x, function.right_inverse (inv_act x) (act x))
(self_distrib : ∀ {x y z : α}, act x (act y z) = act (act x y) (act x z))
```
This would let someone formalize things about shelves more easily, but that's at the cost of the action not automatically being an `equiv` without an additional accessor function.

Reply:
I ended up going with the expanded-out version, and then I added `rack.act : R → (R ≃ R)` to get the equivalence version.  It seems to make things cleaner!